# Scenario 1: Online Shopping Cart
Input:
• The user enters the product name (string).
• The user enters the quantity (integer).

Output:
• The program displays the product name, quantity, and calculated total price (formatted with two decimal plces).


In [39]:
def calculate_total_price(product_name, quantity):
    product_prices = {
        "product1": 10.50,
        "product2": 20.75,
        "product3": 15.00
    }
    if product_name in product_prices:
        total_price = product_prices[product_name] * quantity
        print("Product:", product_name)
        print("Quantity:", quantity)
        print("Total Price: {:.2f}".format(total_price))  # Format total price with two decimal places
    else:
        print("Product not found in the database.")

product_name = input("Enter the product name: ")
quantity = int(input("Enter the quantity: "))

calculate_total_price(product_name, quantity)
      

Enter the product name:  product2
Enter the quantity:  2


Product: product2
Quantity: 2
Total Price: 41.50


# Scenario 2: Restaurant Tip Calculator
Input:
• The user enters the bill amount (float).
• The user selects a tip percentage (integer) from a menu (e.g., 15%, 18%, 20%).

Output:
• The program displays the bill amount, tip amount (formatted with two decimal places), and total bill with tip (formatted with two decimal plces).


In [ ]:
def calculate_tip(bill_amount, tip_percentage):
    tip_amount = bill_amount * (tip_percentage / 100)
    total_bill = bill_amount + tip_amount
    print("Bill Amount: {:.2f}".format(bill_amount))
    print("Tip Amount ({}%): {:.2f}".format(tip_percentage, tip_amount))
    print("Total Bill with Tip: {:.2f}".format(total_bill))

bill_amount = float(input("Enter the bill amount: "))
print("Select tip percentage:")
print("1. 15%")
print("2. 18%")
print("3. 20%")
choice = int(input("Enter your choice (1, 2, or 3): "))

if choice == 1:
    tip_percentage = 15
elif choice == 2:
    tip_percentage = 18
elif choice == 3:
    tip_percentage = 20
else:
    print("Invalid choice. Defaulting to 15%.")
    tip_percentage = 15

calculate_tip(bill_amount, tip_percentage)

# Scenario 3: Event Registration
Input:
• The user enters their first name (string).
• The user enters their last name (string).
• The user enters the number of tickets they want to purchase (integer).

Output:
• The program displays a confirmation message with the user's full name, number of tickets, and total price (formatted based on a ticket price, for example $25).


In [ ]:
def calculate_total_price(ticket_price, num_tickets):
    total_price = ticket_price * num_tickets
    return total_price

first_name = input("Enter your first name: ")
last_name = input("Enter your last name: ")
num_tickets = int(input("Enter the number of tickets you want to purchase: "))

ticket_price = 25

total_price = calculate_total_price(ticket_price, num_tickets)

print("Thank you, {} {}, for your purchase!".format(first_name, last_name))
print("You have purchased {} ticket(s) at ${:.2f} each.".format(num_tickets, ticket_price))
print("Total Price: ${:.2f}".format(total_price)) 

# Dictionary

Scenario 1: Inventory Management
Input:
A dictionary containing product information, where keys are product IDs (strings) and values are dictionaries with details like name (string), price (float), and quantity (integer# ).

Task:
1.	Display all products with their details (name, price, quatity).


In [11]:
def display_product_details(products):
    print("Product Details:")
    print("ID\tName\t\tPrice\tQuantity")
    for product_id, product_info in products.items():
        name = product_info['name']
        price = product_info['price']
        quantity = product_info['quantity']
        print("{}\t{}\t${:.2f}\t{}".format(product_id, name, price, quantity))

products = {
    '001': {'name': 'Product A', 'price': 10.50, 'quantity': 100},
    '002': {'name': 'Product B', 'price': 20.75, 'quantity': 50},
    '003': {'name': 'Product C', 'price': 15.00, 'quantity': 75}
    # Add more products as needed
}

display_product_details(products)

Product Details:
ID	Name		Price	Quantity
001	Product A	$10.50	100
002	Product B	$20.75	50
003	Product C	$15.00	75


# Task:2.	Allow the user to update the quantity of an existing product.

In [10]:
def display_product_details(products):
    print("Product Details:")
    print("ID\tName\t\tPrice\tQuantity")
    for product_id, product_info in products.items():
        name = product_info['name']
        price = product_info['price']
        quantity = product_info['quantity']
        print("{}\t{}\t${:.2f}\t{}".format(product_id, name, price, quantity))

def update_product_quantity(products, product_id, new_quantity):
    if product_id in products:
        products[product_id]['quantity'] = new_quantity
        print("Quantity of product '{}' updated to {}.".format(products[product_id]['name'], new_quantity))
    else:
        print("Product ID '{}' not found.".format(product_id))

products = {
    '001': {'name': 'Product A', 'price': 10.50, 'quantity': 100},
    '002': {'name': 'Product B', 'price': 20.75, 'quantity': 50},
    '003': {'name': 'Product C', 'price': 15.00, 'quantity': 75}
}

display_product_details(products)

product_id = input("Enter the product ID to update quantity: ")
new_quantity = int(input("Enter the new quantity: "))

update_product_quantity(products, product_id, new_quantity)

display_product_details(products)

Product Details:
ID	Name		Price	Quantity
001	Product A	$10.50	100
002	Product B	$20.75	50
003	Product C	$15.00	75


Enter the product ID to update quantity:  003
Enter the new quantity:  500


Quantity of product 'Product C' updated to 500.
Product Details:
ID	Name		Price	Quantity
001	Product A	$10.50	100
002	Product B	$20.75	50
003	Product C	$15.00	500


# Scenario 2: Analysing Book Inventory
Input:
Python
book_inventory = {
 "Moby Dick": {
 "author": "Herman Melville",
 "genre": "Adventure",
 "in_stock": 10
 },
 "Pride and Prejudice": {
 "author": "Jane Austen",
 "genre": "Romance",
 "in_stock": 5
 },
 "The Lord of the Rings": {
 "author": "J.R.R. Tolkien",
 "fantasy": "Epic Fantasy",
 "in_stock": 8 # Typo, missingg# Task:3.re"
 }
}
Task:
1.	Find the total numb in stock.

# Task:3.

In [13]:
book_inventory = {
 "Moby Dick": {
 "author": "Herman Melville",
 "genre": "Adventure",
 "in_stock": 10
 },
 "Pride and Prejudice": {
 "author": "Jane Austen",
 "genre": "Romance",
 "in_stock": 5
 },
 "The Lord of the Rings": {
 "author": "J.R.R. Tolkien",
 "fantasy": "Epic Fantasy",
 "in_stock": 8 # Typo, missing key "genre"
 }
}
total_books_in_stock = sum(book["in_stock"] for book in book_inventory.values())
print("Total number of books in stock:", total_books_in_stock)

Total number of books in stock: 23


# Task:3. Check if a specific book ("The Lord of the Rings") is in stock and how many copies are available.

In [15]:
book_inventory = {
 "Moby Dick": {
 "author": "Herman Melville",
 "genre": "Adventure",
 "in_stock": 10
 },
 "Pride and Prejudice": {
 "author": "Jane Austen",
 "genre": "Romance",
 "in_stock": 5
 },
 "The Lord of the Rings": {
 "author": "J.R.R. Tolkien",
 "fantasy": "Epic Fantasy",
 "in_stock": 8 # Typo, missing key "genre"
 }
}
book_title = "The Lord of the Rings"
if book_title in book_inventory:
    copies_available = book_inventory[book_title].get("in_stock", 0)
    print("{} is in stock with {} copies available.".format(book_title, copies_available))
else:
    print("{} is not found in the inventory.".format(book_title))


The Lord of the Rings is in stock with 8 copies available.


# Task:4.	Correct the typo in "The Lord of the Rings" by adding the missing "genre" key.

In [16]:
print("Updated Book Inventory:")
for book_title, book_info in book_inventory.items():
    print(book_title)
    for key, value in book_info.items():
        print("\t{}: {}".format(key, value))

Updated Book Inventory:
Moby Dick
	author: Herman Melville
	genre: Adventure
	in_stock: 10
Pride and Prejudice
	author: Jane Austen
	genre: Romance
	in_stock: 5
The Lord of the Rings
	author: J.R.R. Tolkien
	fantasy: Epic Fantasy
	in_stock: 8


# Task:5. Display all books with their details in a user-friendly format.

In [17]:
print("Book Inventory:")
for title, details in book_inventory.items():
    print("Title:", title)
    print("\tAuthor:", details.get("author"))
    print("\tGenre:", details.get("genre"))
    print("\tIn Stock:", details.get("in_stock"))


Book Inventory:
Title: Moby Dick
	Author: Herman Melville
	Genre: Adventure
	In Stock: 10
Title: Pride and Prejudice
	Author: Jane Austen
	Genre: Romance
	In Stock: 5
Title: The Lord of the Rings
	Author: J.R.R. Tolkien
	Genre: None
	In Stock: 8


# Scenario 3: Online Shopping Cart
Input:
Python
cart = {
 "apple": 3, # Quantity of each item
 "banana": 2,
 "milk": 1
}
prices = {
 "apple": 1.50,
 "banana": 0.75,
 "milk": 3.00
}
# Task:1.	Calculate the total cost of each item in the cart.

In [19]:
cart = {
 "apple": 3, # Quantity of each item
 "banana": 2,
 "milk": 1
}
prices = {
 "apple": 1.50,
 "banana": 0.75,
 "milk": 3.00
}
total_costs = {}
for item, quantity in cart.items():
    total_costs[item] = quantity * prices.get(item, 0)

print("Total cost of each item in the cart:")
for item, cost in total_costs.items():
    print("{}: ${:.2f}".format(item, cost)) 

Total cost of each item in the cart:
apple: $4.50
banana: $1.50
milk: $3.00


# Task 2.	Compute the overall subtotal for all items before tax.

In [22]:
cart = {
 "apple": 3, # Quantity of each item
 "banana": 2,
 "milk": 1
}
prices = {
 "apple": 1.50,
 "banana": 0.75,
 "milk": 3.00
}
subtotal = sum(cart[item] * prices.get(item, 0) for item in cart)
print("Overall subtotal before tax: ${:.2f}".format(subtotal))

Overall subtotal before tax: $9.00


# Task 3.	Add a 10% sales tax and calculate the final total.
Apple: $4.50
Banana: $1.50
Milk: $3.00

In [31]:
item_totals = {item: quantity * price for item, quantity in cart.items() for item, price in prices.items()}
item_totals = {item: cart[item] * prices[item] for item in cart}
item_totals = {item: cart[item] * prices[item] for item in cart}
#print (item_totals)

tax_rate = 0.10
item_tax = {item: item_cost * tax_rate for item, item_cost in item_totals.items()}
item_final_total = {item: item_cost + tax for item, item_cost in item_totals.items() for item, tax in item_tax.items()}
for item, total in item_final_total.items():
    print("{}: ${:.2f}".format(item.capitalize(), total))

Apple: $3.45
Banana: $3.15
Milk: $3.30


# List comprehension!

Scenario 1: Filtering Available Hotel Rooms
• Input: A list of dictionaries representing hotel rooms, where each dictionary has keys
like room_type (e.g., "Single", "Double") and availability (e.g., True, False).
• Task: Create a new list containing only the room dictionaries with availability set to
True.


In [32]:
rooms = [
    {"room_type": "Single", "availability": True},
    {"room_type": "Double", "availability": True},
    {"room_type": "Suite", "availability": False},
    {"room_type": "Twin", "availability": True},
    {"room_type": "Deluxe", "availability": False}
]

available_rooms = [room for room in rooms if room["availability"]]
print("Available Rooms:")
for room in available_rooms:
    print(room)

Available Rooms:
{'room_type': 'Single', 'availability': True}
{'room_type': 'Double', 'availability': True}
{'room_type': 'Twin', 'availability': True}


# Scenario 2: Extracting High Scores from a Game
• Input: A list of player scores (e.g., player_scores = [80, 95, 72, 100, 88]) and a
threshold score (e.g., threshold = 90).
• Task: Create a new list containing only the scores that are greater than or equal to the
threshld.


In [33]:
player_scores = [80, 95, 72, 100, 88]
threshold = 90
high_scores = [score for score in player_scores if score >= threshold]
print("High Scores:", high_scores)

High Scores: [95, 100]


# Scenario 3: Building a Shopping Cart with Conditional Discounts
• Input: A list of product dictionaries with keys like name (e.g., "T-Shirt"), price (e.g.,
20), and category (e.g., "Clothing"), and a list of discount rules represented as
dictionaries with keys like category (e.g., "Clothing") and discount_percent (e.g.,
10).
• Task: Create a new list containing product dictionaries with their final prices after
applying any applicable discunts.


In [34]:
def apply_discount(product, discount_rules):
    for rule in discount_rules:
        if product["category"] == rule["category"]:
            discount_percent = rule["discount_percent"]
            product["price"] *= (1 - discount_percent / 100)
    return product

products = [
    {"name": "T-Shirt", "price": 20, "category": "Clothing"},
    {"name": "Jeans", "price": 50, "category": "Clothing"},
    {"name": "Sneakers", "price": 80, "category": "Shoes"}
]

discount_rules = [
    {"category": "Clothing", "discount_percent": 10},
    {"category": "Shoes", "discount_percent": 15}
]

final_prices = [apply_discount(product, discount_rules) for product in products]
print("Product\t\tPrice\tFinal Price")
for product in final_prices:
    print("{}\t\t${:.2f}\t${:.2f}".format(product["name"], product["price"], product["price"])) 

Product		Price	Final Price
T-Shirt		$18.00	$18.00
Jeans		$45.00	$45.00
Sneakers		$68.00	$68.00


# Nested list comprehension
# 
Scenario 1: Generating Combinations:
• Task: Create a list of all possible combinations of two toppings on a pizza (e.g., from
"pepperoni", "mushrooms", "onions").
• Input: Lists of topping options (e.g., toppings1 = ["pepperoni", "mushrooms"],
toppings2 = ["onions", "pepprs"])


In [35]:
def pizza_combinations(toppings1, toppings2):
    combinations = []
    for topping1 in toppings1:
        for topping2 in toppings2:
            combinations.append([topping1, topping2])
    return combinations

toppings1 = ["pepperoni", "mushrooms"]
toppings2 = ["onions", "peppers"]

all_combinations = pizza_combinations(toppings1, toppings2)
for combination in all_combinations:
    print(combination)

['pepperoni', 'onions']
['pepperoni', 'peppers']
['mushrooms', 'onions']
['mushrooms', 'peppers']


# Scenario 2: Movie Recommendation System (Considering User Preferences)
• Task: Given a list of movies and their genres (movies = [{"title": "Baahubali",
"genre": "Action"}, ...]), create a dictionary mapping genres to movie titles based
on user preferences (e.g., preferred genre entered as user_pref="Comedy").

In [38]:
def map_movies_to_genre(movies, user_pref):
    genre_to_movies = {}
    for movie in movies:
        title = movie["title"]
        genre = movie["genre"]
        if genre == user_pref:
            if genre not in genre_to_movies:
                genre_to_movies[genre] = [title]
            else:
                genre_to_movies[genre].append(title)
    return genre_to_movies

movies = [
    {"title": "Baahubali", "genre": "Action"},
    {"title": "Dumb and Dumber", "genre": "Comedy"},
    {"title": "The Shawshank Redemption", "genre": "Drama"},
    {"title": "Deadpool", "genre": "Action"},
    {"title": "The Hangover", "genre": "Comedy"},
    {"title": "Inception", "genre": "Sci-Fi"}
]

user_pref = "Comedy"

genre_to_movies = map_movies_to_genre(movies, user_pref)
print(genre_to_movies)

{'Comedy': ['Dumb and Dumber', 'The Hangover']}
